Sources/LoopAlgorithm/Insulin/DoseMath.swift

```diff
@@ -0,0 +1,339 @@
+        var eventualGlucose: GlucoseValue!
+        var correctingGlucose: GlucoseValue?
+        var minCorrectionUnits: Double?
+        var effectedSensitivityAtMinGlucose: Double?
+
+        let unit = HKUnit.milligramsPerDeciliter
+
+        guard self.count > 0 else {
+            preconditionFailure("Unable to compute correction for empty glucose array")
+        }
+
+        // If this is not true, then this method will return very large doses. For example, it takes a *lot* of
+        // insulin to bring a predicted glucose of 200 down to range within 30 minutes, so we assert that the
+        // prediction includes values out to the end of insulin activity.
+        guard self.last!.startDate >= date.addingTimeInterval(model.effectDuration) else {
+            preconditionFailure("Minimization method requires that glucose prediction covers at least the insulin effect duration.")
+        }
+
+        let suspendThresholdValue = suspendThreshold.doubleValue(for: unit)
+
+        // For each prediction above target, determine the amount of insulin necessary to correct glucose based on the modeled effectiveness of the insulin at that time
+        for prediction in self {
+            guard prediction.startDate >= date else {
+                continue
+            }
+
+            // If any predicted value is below the suspend threshold, return immediately
+            guard prediction.quantity >= suspendThreshold else {
+                return .suspend(min: prediction)
+            }
+
+            eventualGlucose = prediction
+
+            let predictedGlucoseValue = prediction.quantity.doubleValue(for: unit)
+            let time = prediction.startDate.timeIntervalSince(date)
+
+            guard let correctionRangeItem = correctionRange.closestPrior(to: prediction.startDate) else {
+                preconditionFailure("Correction range must cover date: \(prediction.startDate)")
+            }
+
+            // Compute the target value as a function of time since the dose started
+            let targetValue = targetGlucoseValue(
+                percentEffectDuration: time / model.effectDuration,
+                minValue: suspendThresholdValue,
+                maxValue: correctionRangeItem.value.averageValue(for: unit)
+            )
+
+            // Compute the dose required to bring this prediction to target:
+            // dose = (Glucose Δ) / (% effect × sensitivity)
+
+            let isfSegments = insulinSensitivity.filterDateRange(date, prediction.startDate)
+
+            var isfEnd: TimeInterval?
+
+            let effectedSensitivity = isfSegments.reduce(0) { partialResult, segment in
+                let start = Swift.max(date, segment.startDate).timeIntervalSince(date)
+                let end = Swift.min(prediction.startDate, segment.endDate).timeIntervalSince(date)
+                let percentEffected = model.percentEffectRemaining(at: start) - model.percentEffectRemaining(at: end)
```

Comment:
What is the temporal granularity for the start/end calculations? Seconds?

Reply:
Double precision. Do you see an issue?

Reply:
No issue, was just confirming for android port and currently can only access here:)

Reply:
Sorry, there is one potential modification to consider. The removal of `isfEnd` and validation of the coverage of the timeline as a separate step, using the actual end dates of the segments instead of an accumulator. 

`guard let lastSegmentEnd = insulinSensitivity.max(by: { $0.endDate < $1.endDate })?.endDate, lastSegmentEnd >= prediction.startDate else {
                preconditionFailure("Sensitivity timeline must cover date: \(prediction.startDate)")
            }`

Reply:
Could you show a code version of what you're suggesting?

Reply:
Merging this since approved, but we can continue the discussion.

Reply:
Sure, can do on a future PR